src/commandargs.cpp

```diff
@@ -8,42 +8,65 @@ static Setting* CurrentState = nullptr;
 
 bool MatchesFlag(String token, Setting** matchedSetting)
 {
+    if(token.length() < 1)
+    {
+        return false;
+    }
+
     for(Setting& s: *Config)
     {
         if(token == s.Flag)
         {
+            CurrentState = &s;
             *matchedSetting = &s;
             return true;
         }
     }
+    
+    // If an unknown which resembles a flag is passed, inform the user
+    if(token.at(0) == '-')
+    {
+        Usage(std::vector<SettingOption>());
+    }
+    
     return false;
 }
 
-void ParseCommandArgs(int argc, char* argv[], ProgramConfiguration config)
+void ParseCommandArgs(int argc, char* argv[], ProgramConfiguration* config)
 {
-    Config = &config;
+    Config = config;
     CurrentState = nullptr;
+    bool inFlag = false;
 
     std::vector<SettingOption> CurrentSettingOptions;
     CurrentSettingOptions.reserve(8);
 
     for(int i=1; i<argc; i++)
     {
         String arg = argv[i];
+        
         Setting* newSetting = nullptr;
         if(MatchesFlag(arg, &newSetting))
-        {
-            if(CurrentState != nullptr)
+        {            
+            if(CurrentState->Action(CurrentSettingOptions))
```

Comment:
I don't see any segfaulting behavior on my end. 

The CurrentState is populated by the flag matching function guaranteed iirc. 

Testing on my end:

```
seh@arrietty:~/repos/pebl-sean$ ./pebble.exe 
################################################################################
2
################################################################################
seh@arrietty:~/repos/pebl-sean$ 
seh@arrietty:~/repos/pebl-sean$ ./pebble.exe program.pebl 
################################################################################
2
################################################################################
seh@arrietty:~/repos/pebl-sean$ ./pebble.exe --log sev1 program.pebl 
################################################################################
2
################################################################################
seh@arrietty:~/repos/pebl-sean$ ./pebble.exe --log sev1 
################################################################################
2
################################################################################
seh@arrietty:~/repos/pebl-sean$ 
```

Reply:
Here's what I'm getting?
```
PS C:\Users\t-ket\Repos\Chief> gdb --args ./pebble.exe program.pebl
GNU gdb (GDB) 8.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "i686-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./pebble.exe...done.
(gdb) run
Starting program: C:\Users\t-ket\Repos\Chief\pebble.exe program.pebl
[New Thread 11132.0x5eb0]
[New Thread 11132.0x546c]
[New Thread 11132.0x5fe4]
[New Thread 11132.0x49e0]

Thread 1 received signal SIGSEGV, Segmentation fault.
0x00000000 in ?? ()
(gdb) backtrace
#0  0x00000000 in ?? ()
#1  0x0041c87d in ParseCommandArgs (argc=2, argv=0x1691700, config=0x467194 <Config>) at src/commandargs.cpp:51
#2  0x00413148 in main (argc=2, argv=0x1691700) at src/main.cpp:82
(gdb) 
```


**RCA:**

The configuration we're using has the Setting with `Flag` `program.pebl` and an `Action = nullptr`
```cpp
ProgramConfiguration Config
{
    {
        /* 
            This is the .pebl file to ingest. 
            The flag string will be filled retro-actively with the program name, if any. 
            Must be the 0'th Setting. 
            Note: making Flag a vector rather than string could allow multiple files/args
        */
        "Pebl File", "program.pebl", nullptr
    },
```

Running `$ ./pebble.exe program.pebl` will trigger `MatchesFlag` to return true because `program.pebl` token matches the `Flag` of the first `Setting` aka `Config.At(0)`
```cpp
bool MatchesFlag(String token, Setting** matchedSetting)
...
    for(Setting& s: *Config)
    {
        if(token == s.Flag)
        {
            CurrentState = &s;
            *matchedSetting = &s;
            return true;
        }
    }


```

The method sets `CurrentState` to `Config.at(0)` which has `Action = nullptr`

```cpp
void ParseCommandArgs(int argc, char* argv[], ProgramConfiguration* config)
...
        if(MatchesFlag(arg, &newSetting))
        {            
            if(CurrentState->Action(CurrentSettingOptions))
            {
                inFlag = true;
            }
...
```
Dereferencing `CurrentState->Action` causes a segfault here.

I think the logic works... I'm really confused as to why you aren't seeing it. Maybe force `make` your `pebble.exe` build?

Reply:
Resolved by 1100bf4

I was using the wrong branch 🙃